dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTinyVectorIndexWriter.cpp

```diff
@@ -0,0 +1,234 @@
+
+
+namespace DB::ErrorCodes
+{
+extern const int ABORTED;
+} // namespace DB::ErrorCodes
+
+namespace DB::DM
+{
+
+ColumnFileTinyVectorIndexWriter::LocalIndexBuildInfo ColumnFileTinyVectorIndexWriter::getLocalIndexBuildInfo(
+    const LocalIndexInfosSnapshot & index_infos,
+    const ColumnFilePersistedSetPtr & file_set)
+{
+    assert(index_infos != nullptr);
+
+    LocalIndexBuildInfo build;
+    build.indexes_to_build = std::make_shared<LocalIndexInfos>();
+    build.file_ids.reserve(file_set->getColumnFileCount());
+    for (const auto & file : file_set->getFiles())
+    {
+        auto * tiny_file = file->tryToTinyFile();
+        if (!tiny_file)
+            continue;
+
+        bool any_new_index_build = false;
+        for (const auto & index : *index_infos)
+        {
+            auto schema = tiny_file->getSchema();
+            assert(schema != nullptr);
+            // The ColumnFileTiny may be built before col_id is added. Skip build indexes for it
+            if (!schema->getColIdToOffset().contains(index.column_id))
+                continue;
+
+            // Index already built, skip
+            if (tiny_file->hasIndex(index.index_id))
+                continue;
+
+            any_new_index_build = true;
+            // FIXME: the memory usage is not accurate, but it's fine for now.
+            build.estimated_memory_bytes += tiny_file->getBytes();
+
+            // Avoid duplicate index build
+            if (std::find(build.index_ids.begin(), build.index_ids.end(), index.index_id) == build.index_ids.end())
+            {
+                build.indexes_to_build->emplace_back(index);
+                build.index_ids.emplace_back(index.index_id);
+            }
+        }
+
+        if (any_new_index_build)
+        {
+            build.file_ids.emplace_back(LocalIndexerScheduler::ColumnFileTinyID(tiny_file->getDataPageId()));
+        }
+    }
+
+    build.file_ids.shrink_to_fit();
+    return build;
+}
+
+ColumnFileTinyPtr ColumnFileTinyVectorIndexWriter::buildIndexForFile(
+    const ColumnDefines & column_defines,
+    const ColumnDefine & del_cd,
+    const ColumnFileTiny * file,
+    ProceedCheckFn should_proceed) const
+{
+    // read_columns are: DEL_MARK, COL_A, COL_B, ...
+    // index_builders are: COL_A, COL_B, ...
+
+    ColumnDefinesPtr read_columns = std::make_shared<ColumnDefines>();
+    read_columns->reserve(options.index_infos->size() + 1);
+    read_columns->push_back(del_cd);
+
+    std::unordered_map<ColId, std::vector<VectorIndexBuilderPtr>> index_builders;
+
+    std::unordered_map<ColId, std::vector<LocalIndexInfo>> col_indexes;
+    for (const auto & index_info : *options.index_infos)
+    {
+        if (index_info.type != IndexType::Vector)
+            continue;
+        col_indexes[index_info.column_id].emplace_back(index_info);
+    }
+
+    for (const auto & [col_id, index_infos] : col_indexes)
+    {
+        // Make sure the column_id is in the schema.
```

Comment:
So we are sure that the schema is up-to-date here? I assume there is somewhere that we call syncSchema or something to ensure this?

Reply:
No, we do not need to guarantee this. If one file's schema is not up-to-date, we will just skip the new created index.